samples/bluetooth/unicast_audio_client/src/main.c

```diff
@@ -38,39 +38,151 @@ static K_SEM_DEFINE(sem_stream_qos, 0, 1);
+ *
+ * @param buf Destination buffer
+ * @param length_us Length of the buffer in microseconds
+ * @param frequency_hz frequency in Hz
+ * @param sample_rate_hz sample-rate in Hz.
+ */
+static void fill_audio_buf_sin(int16_t *buf, int length_us, int frequency_hz, int sample_rate_hz)
 {
-	while (len-- != 0) {
-		printk("%02x", *ptr++);
+	const int sine_period = sample_rate_hz / frequency_hz;
+	const int num_samples = (length_us * sample_rate_hz) / 1000000;
+	const float step = 3.1415 / sine_period;
+
+	for (int i = 0; i < num_samples; i++) {
+		float sample = sin(i * step);
+
+		buf[i] = (int16_t)(INT16_MAX * sample);
 	}
 }
 
-static void print_codec(const struct bt_codec *codec)
+static const char *get_err_str(uint16_t err)
 {
-	printk("codec 0x%02x cid 0x%04x vid 0x%04x count %u\n",
-	       codec->id, codec->cid, codec->vid, codec->data_count);
+	switch (err) {
+	case LC3_ENCODE_ERR_FREE:
+		return "LC3_ENCODE_ERR_FREE";
+	case LC3_ENCODE_ERR_INVALID_CONFIGURATION:
+		return "LC3_ENCODE_ERR_INVALID_CONFIGURATION";
+	case LC3_ENCODE_ERR_INVALID_BYTE_COUNT:
+		return "LC3_ENCODE_ERR_INVALID_BYTE_COUNT";
+	case LC3_ENCODE_ERR_INVALID_BITS_PER_AUDIO_SAMPLE:
+		return "LC3_ENCODE_ERR_INVALID_BITS_PER_AUDIO_SAMPLE";
+	case LC3_ENCODE_ERR_ENCODER_ALLOCATION:
+		return "LC3_ENCODE_ERR_ENCODER_ALLOCATION";
+	}
+	return "UNKNOWN LC3 ENCODE ERROR";
+}
 
-	for (size_t i = 0; i < codec->data_count; i++) {
-		printk("data #%zu: type 0x%02x len %u\n",
-		       i, codec->data[i].data.type,
-		       codec->data[i].data.data_len);
-		print_hex(codec->data[i].data.data,
-			  codec->data[i].data.data_len -
-				sizeof(codec->data[i].data.type));
-		printk("\n");
+static void lc3_audio_timer_timeout(struct k_work *work)
+{
+	/* For the first call-back we push multiple audio frames to the buffer to use the
+	 * controller ISO buffer to handle jitter.
+	 */
+	const uint8_t prime_count = 2;
+	static int64_t start_time;
+	static int32_t pdu_cnt;
+	int32_t pdu_goal_cnt;
+	int64_t uptime, run_time_ms;
+	int ret;
+	struct net_buf *buf;
+	uint8_t *net_buffer;
+	uint16_t lc3_ret;
+	const uint16_t tx_sdu_len = audio_stream.iso->qos->tx->sdu;
+
+	k_work_schedule(&audio_send_work, K_MSEC(10));
+
+	if (start_time == 0) {
+		/* Read start time and produce the number of frames needed to catch up with any
+		 * inaccuracities in the timer. by calculating the number of frames we should
+		 * have send and compare to how many wa have actually sent.
+		 */
+		start_time = k_uptime_get();
 	}
 
-	for (size_t i = 0; i < codec->meta_count; i++) {
-		printk("meta #%zu: type 0x%02x len %u\n",
-		       i, codec->meta[i].data.type,
-		       codec->meta[i].data.data_len);
-		print_hex(codec->meta[i].data.data,
-			  codec->meta[i].data.data_len -
-				sizeof(codec->meta[i].data.type));
-		printk("\n");
+	uptime = k_uptime_get();
+	run_time_ms = uptime - start_time;
+	pdu_goal_cnt = ((run_time_ms * 10) / (AUDIO_LENGTH_US / 100)); /* scale to allow 7.5ms */
```

Comment:
The point with this scaling is that if one uses 7500us and the division was done at /1000 it would be truncated to 7.

Reply:
Might be the lack of coffee talking, but I fail to see how this fixes that. Doesn't this give the same result a `run_time_ms /  (AUDIO_LENGTH_US / 1000)`?

Reply:
As it uses fixed point calculations 7500/1000 becomes 7, hence if you divide the run-time by 7 you will be 0.5ms off for each SDU interval and data will be transmitted too fast.
(AUDIO_LENGTH_US / 100) will be evaluated at compile time.

Reply:
So the implementation uses 10ms by default, but I want to code to allow selection of 7.5ms :-)

Reply:
> As it uses fixed point calculations 7500/1000 becomes 7, hence if you divide the run-time by 7 you will be 0.5ms off for each SDU interval and data will be transmitted too fast. (AUDIO_LENGTH_US / 100) will be evaluated at compile time.

And the current calculation `7500 * 10 / (1000000 / 100)` gives `(7500 * 10) / 10000` == `75000 / 10000` (as you basically just multiple both sides by 10), but doesn't that also just give 7? Or am I missing something here?

Reply:
It is the other way around, say you set AUDIO_LENGTH_US  to 7500 (the only alternative to 10ms). Then you run for 30s:
(30.000*10)/(7.500/100) = 4.000
But if we do it
30.000/(7.500/1.000) => 30.000/7 => 4.285

Hence wee transmit 285 additional audio frames in 30s. In other words we transmit data ~7% too fast

Reply:
Ah, I had misunderstand the values :) That makes sense

Reply:
Comment is unclear, and I find the mixing of two operations (computation and scaling) in one calculation confusing.
How about just casting the numbers to floats before the calculation, and then the result back to int again?

`pdu_goal_cnt = (int32_t)( (double)run_time_ms / (  (double)AUDIO_LENGTH_US / 1000) )`

Reply:
@asbjornsabo thanks for the suggestion. That would work but be quite a heavy operation on a CPU without HW floating point support. Likewise if run_time was stored in microseconds it would require a 64bit variable to support a decent run-time - introducing 64 bit math operation not all CPU's support. I'll have a second look at it to see if there is an alternative solution and at least improve the comment.